### PR TITLE
feat(gateway): serve the Anthropic Messages API for Claude Code

### PR DESCRIPTION
**Pull request by rohoswagger**
## Description

Adds the Anthropic Messages API to the Onyx LLM gateway so Claude Code (and other Anthropic-native clients) can use it via `ANTHROPIC_BASE_URL`:

- `POST /gateway/v1/messages` — streaming (Anthropic SSE `event:` framing: `message_start`/`ping`/`content_block_*`/`message_delta`/`message_stop`, in-band `error` events) and non-streaming.
- `POST /gateway/v1/messages/count_tokens` — Claude Code calls this for context tracking; falls back to a rough estimate rather than failing.

Design, following the Responses-API precedent in this stack:
- Input translation reuses LiteLLM's Anthropic→OpenAI adapter; system/tool list-content is collapsed to fit our str-only message fields; tools are hand-mapped (`input_schema` → function `parameters`).
- Output is hand-rolled frozen `_WireModel` payloads. Text streams live; `tool_use` blocks emit at end-of-stream as one `input_json_delta` (same tradeoff as the Responses endpoint).
- Usage converts back to Anthropic accounting: `input_tokens` excludes cache reads/writes (LiteLLM folds them into `prompt_tokens`).
- Both Anthropic thinking APIs are supported off a single `ReasoningEffort`: legacy `thinking.type=enabled` + `budget_tokens` and adaptive `thinking.type=adaptive` + `output_config.effort`; the LLM layer re-derives the right API per model.
- Refused (can't be honored): server-side/hosted tools, named-tool `tool_choice`, multimodal `tool_result` content. Tolerated-and-ignored (dropping only shapes output): `stop_sequences`, `top_p`/`top_k`, `metadata`.
- Thinking blocks replayed in history are stripped (signatures can't round-trip through the internal abstraction) — deliberate, matches existing chat-path behavior.

Auth is the existing gateway model: Bearer PAT with `USE_LLM_GATEWAY` scope (`ANTHROPIC_AUTH_TOKEN`; `ANTHROPIC_API_KEY`/x-api-key is not supported), model id `<provider_id>/<model_name>` via `ANTHROPIC_MODEL`.

## How Has This Been Tested?

- 52 new unit tests in `test_anthropic_messages_api.py` (conversion, refusals, thinking/effort matrix, stop reasons, non-streaming shape, streaming event order/framing, stream errors, usage math incl. cache subtraction, auth gating, count_tokens) — 129 pass in the gateway suite.
- Live end-to-end against a local instance with a real Anthropic provider: curl (non-streaming, streaming tool-use, count_tokens) and real Claude Code CLI turns through the gateway — plain text turn and a 4-turn agentic tool-use session (`ls` + file read), including prompt-cache hits (cache_read ≈ full prefix on later turns) and correct cost attribution in the usage ledger.
- 4-lens review pass (correctness / security / tests / regressions): no critical/high findings; regressions lens verified the diff is purely additive.

### Known minor items (for the reviewer, deliberately not fixed here)

- Precedence when a client sends BOTH legacy `thinking.budget_tokens` and `output_config.effort`: `output_config.effort` wins. Debatable; real clients target one API per model generation, so the combination should be rare.
- `count_tokens` runs the tokenizer without a token-budget check (nothing is generated); an authorized user could burn CPU on it. Body-size limits still apply.
- Test gaps flagged by review: no per-route single-permission-dependency tests for the two new routes (the blanket router auth test does cover them), no test for the usage clamp actually clamping, none for the count_tokens fallback branch or the hosted-tool identifier fallback when `name` is absent.

## Additional Options

- [ ] [Optional] Please cherry-pick this PR to the latest release version.
- [x] [Optional] Override Linear Check

---

**Stack:**
1. [gateway-reject-unsupported-params #13650](https://github.com/onyx-dot-app/onyx/pull/13650)

---
## Summary by cubic
Adds Anthropic Messages API support in the LLM gateway so Claude Code and Anthropic-native clients can use `ANTHROPIC_BASE_URL`. Includes SSE streaming and token counting with Anthropic-accurate usage and stop reasons, plus a small stream-loop refactor for clearer cancellation and completion.

- **New Features**
  - Added `POST /gateway/v1/messages` (streaming SSE: `message_start`/`ping`/`content_block_*`/`message_delta`/`message_stop`, and non-streaming) and `POST /gateway/v1/messages/count_tokens` (falls back to a rough estimate).
  - Translates Anthropic requests via `litellm`; collapses system/tool content to strings; maps tools (`input_schema` → function `parameters`); rejects multimodal `tool_result`.
  - Supports both thinking APIs via one `ReasoningEffort`; converts usage to Anthropic accounting (`input_tokens` excludes cache reads/writes). Ignores `stop_sequences` and sampling params.

- **Bug Fixes**
  - Correct stop reason mapping: report `max_tokens` over `tool_use` when a tool-call turn truncates.
  - Rejects hosted/server-side tools and named-tool `tool_choice`; validates `tool_choice` values.
  - Streaming: emit a single in-band `error` and close upstream; replace for-else loops with early returns to improve cancellation and final-frame handling, with tighter typing for stream events and the adapter.

<sup>Written for commit 07cbb781cecf35283b0145a8fb62b1e4b9500e88. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/onyx-dot-app/onyx/pull/13687?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: backend/onyx/server/gateway/api.py [hunk 1]
```diff
@@ -951,3 +1411,59 @@ def gateway_responses(
     if isinstance(result, StreamingResponse):
         return result
     return JSONResponse(content=result.to_wire())
+
+
+@router.post("/v1/messages")
```
[the remark this answers is not in the document]
Comment (rohoswagger) [reply backend/onyx/server/gateway/api.py:1438]: Not specific to this PR — these routes sit on the same `/gateway` router as the already-shipped `/v1/chat/completions`, `/v1/responses`, and `/v1/models` endpoints. Clients are pointed at the gateway through the API prefix (`ANTHROPIC_BASE_URL=<host>/api/gateway`), which nginx routes to the API server in Helm/compose deployments — same contract as the existing Codex/opencode integrations (verified live: Claude Code end-to-end via `/api/gateway/v1/messages`). Bare `/gateway/...` was never a supported path for any gateway endpoint.

File: backend/onyx/server/gateway/models.py [hunk 2]
```diff
@@ -510,3 +516,236 @@ class ResponsesFailedEvent(_WireModel):
...
+    input: dict[str, Any]
+
+    @classmethod
+    def create(
+        cls, *, id: str, name: str, input: dict[str, Any]
+    ) -> "AnthropicToolUseBlock":
+        return cls(type="tool_use", id=id, name=name, input=input)
+
+
+AnthropicContentBlock: TypeAlias = AnthropicTextBlock | AnthropicToolUseBlock
```
[the remark this answers is not in the document]
Comment (rohoswagger) [reply backend/onyx/server/gateway/models.py:606]: Deliberate, documented degradation (see PR description): Onyx's internal LLM abstraction carries reasoning as a flat string with no signature field, so real Anthropic thinking blocks can't round-trip — and emitting synthesized blocks with invalid signatures would break clients that replay them (Anthropic rejects tampered signatures). Thinking is still honored on the request side (`thinking`/`output_config.effort` map to ReasoningEffort, so the model does reason); only the block passthrough is stripped, matching the existing chat-path behavior (multi_llm.py drops thinking blocks for the same reason). Verified live: Claude Code operates correctly against this route with thinking-capable models. Signature-preserving passthrough needs a content-block abstraction in onyx/llm/* — out of scope here.

File: backend/onyx/server/gateway/api.py [hunk 3]
```diff
@@ -929,6 +959,422 @@ def handle_responses_request(
...
+    "length": "max_tokens",
+    "tool_calls": "tool_use",
+    "content_filter": "refusal",
+}
+
+
+def _anthropic_stop_reason(finish_reason: str | None, has_tool_use: bool) -> str:
+    # Tool calls dominate: providers routinely report finish_reason "stop" on
+    # tool-call turns.
+    if has_tool_use:
```
Comment (evan-onyx) [inline backend/onyx/server/gateway/api.py:1113]: ?? if it has tool use we just unconditionally say it stopped bc of tool use? seems sus

File: backend/onyx/server/gateway/api.py [hunk 4]
```diff
@@ -929,6 +959,422 @@ def handle_responses_request(
...
+                    if not emit(
+                        AnthropicContentBlockDeltaEvent.create(
+                            index=0,
+                            delta=AnthropicTextDelta.create(
+                                text=chunk.choice.delta.content
+                            ),
+                        )
+                    ):
+                        break
+            else:
```
Comment (evan-onyx) [inline backend/onyx/server/gateway/api.py]: an intentional for-else?! brief comment explaining why would be v good

Comment (evan-onyx) [reply backend/onyx/server/gateway/api.py]: it does look like you could just return instead of break above, and then not put this in an else

**Review by evan-onyx — APPROVED** [review]
some lil nits, feel free to address in follow up(s)

File: backend/onyx/server/gateway/api.py (hunk 3 above)
Comment (rohoswagger) [reply backend/onyx/server/gateway/api.py:1113]: Legit — fixed in 30393c519d. Truncation now takes precedence: `length` maps to `max_tokens` even when tool calls were accumulated (matches Anthropic's own API — a generation cut off mid-tool-call must not look like a completed tool_use turn). The remaining tool-use-wins rule only overrides `stop`/default, which is deliberate: providers routinely report finish_reason "stop" on tool-call turns (comment in code), and reporting end_turn there would make clients treat the turn as final and drop the tool calls. New parametrize row covers length+tool-calls.

File: backend/onyx/server/gateway/api.py (hunk 4 above)
Comment (rohoswagger) [reply backend/onyx/server/gateway/api.py]: Added the comment in 30393c519d — the else runs only when the stream exhausted cleanly; any break (cancellation, failed emit) must skip the completion frames and go straight to the guard's teardown. On restructuring to early returns: I'd rather not — the two sibling stream workers in this file (`_stream_worker`, `_responses_stream_worker`, both merged) use the identical for-else shape, and having one of three parallel workers diverge makes them harder to diff-read; behavior is equivalent either way since returns inside the guard still run its teardown. Happy to change all three uniformly in a follow-up if you feel strongly.

Comment (rohoswagger) [reply backend/onyx/server/gateway/api.py]: Done in 07cbb78 — went with your suggestion and applied it uniformly: all three stream workers (`_stream_worker`, `_responses_stream_worker`, `_anthropic_stream_worker`) now use early `return`s at the cancellation/failed-emit sites, with the completion emission as plain post-loop code — no more for-else. Behavior is identical (returns inside the guard's `with` still run its teardown), and the streaming tests pin the full event sequences across all three.
